File: src/main/scala/com/dyonovan/neotech/pipes/tiles/IPipe.java
package com.dyonovan.neotech.pipes.tiles;

import com.dyonovan.neotech.pipes.network.PipeInformation;
import com.dyonovan.neotech.pipes.network.ResourceEntity;
import net.minecraft.util.EnumFacing;

/**
 * This file was created for NeoTech
 *
 * NeoTech is licensed under the
 * Creative Commons Attribution-NonCommercial-ShareAlike 4.0 International License:
 * http://creativecommons.org/licenses/by-nc-sa/4.0/
 *
 * @author Paul Davis pauljoda
 * @since August 15, 2015
 */
public interface IPipe {

    /**
     * Can the pipe connect on this end?
     * @param facing The face of us
     * @return True if should connect
     */
    boolean canConnect(EnumFacing facing);

    /**
     * Called when a resource enters this pipe
     * @param resource The resource entering
     */
    void onResourceEnteredPipe(ResourceEntity resource);

    /**
     * Used to get the information for this pipe
     */
    PipeInformation getInformation();

    /**
     * Used to get the current position as a long
     * @return BlockPos as long
     */
    Long getPosAsLong();

    /**
     * Used to tell if this can receive a resource
     * @param resource The resource to check, pass null to check if is sink
     * @return True if it can accept, if null is passed and this is a sink, still return true
     */
    boolean canAcceptResource(ResourceEntity resource);
}

File: src/main/scala/com/dyonovan/neotech/pipes/tiles/IPipe.scala
package com.dyonovan.neotech.pipes.tiles

import com.dyonovan.neotech.pipes.network.{PipeInformation, ResourceEntity}
import net.minecraft.tileentity.TileEntity
import net.minecraft.util.EnumFacing

/**
 * This file was created for NeoTech
 *
 * NeoTech is licensed under the
 * Creative Commons Attribution-NonCommercial-ShareAlike 4.0 International License:
 * http://creativecommons.org/licenses/by-nc-sa/4.0/
 *
 * @author Paul Davis pauljoda
 * @since August 16, 2015
 */
trait IPipe extends TileEntity {
    def canConnect(facing: EnumFacing): Boolean = getWorld.getTileEntity(getPos.offset(facing)).isInstanceOf[IPipe]

    def onResourceEnteredPipe(resource: ResourceEntity[_]): Unit = {}

    def getInformation: PipeInformation = {
        val connections = Array[Boolean](false, false, false, false, false, false)
        for(dir <- EnumFacing.values()) {
            getWorld.getTileEntity(getPos.offset(dir)) match {
                case otherPipe: IPipe =>
                    connections(dir.ordinal()) = true
                case _ =>
            }
        }
        new PipeInformation(connections, canAcceptResource(null))
    }

    def canAcceptResource(resource: ResourceEntity[_]): Boolean = false

    def getPosAsLong: Long = getPos.toLong
}
